test/unit/foreman/renderer/scope/macros/base_test.rb

```diff
@@ -92,16 +92,20 @@ class BaseMacrosTest < ActiveSupport::TestCase
 
     test 'should return kernel release and based on backup facts even if there are multiple other facts' do
       host = FactoryBot.create(:host)
+      rhsm_host = FactoryBot.create(:host)
       ansible_kernel_fact = FactoryBot.create(:fact_name, name: 'ansible_kernel', :type => 'FactName::Ansible')
       chef_kernel_fact = FactoryBot.create(:fact_name, name: 'kernel::release', :type => 'FactName::Chef')
       unrelated_fact = FactoryBot.create(:fact_name, name: 'os')
       puppet_and_salt_fact = FactoryBot.create(:fact_name, name: 'kernelrelease')
+      rhsm_fact = FactoryBot.create(:fact_name, name: 'uname::release', :type => 'Katello::RhsmFactName')
       FactoryBot.create(:fact_value, fact_name: ansible_kernel_fact, host: host, :value => '1.2.3')
       FactoryBot.create(:fact_value, fact_name: chef_kernel_fact, host: host, :value => '2.2.2')
       FactoryBot.create(:fact_value, fact_name: unrelated_fact, host: host, :value => 'Fedora 29')
       assert_equal '1.2.3', @scope.host_kernel_release(host)
       FactoryBot.create(:fact_value, fact_name: puppet_and_salt_fact, host: host, :value => '4.5.6')
       assert_equal '4.5.6', @scope.host_kernel_release(host.reload)
+      FactoryBot.create(:fact_value, fact_name: rhsm_fact, host: rhsm_host, :value => '7.8.9')
+      assert_equal '7.8.9', @scope.host_kernel_release(rhsm_host)
```

Comment:
I think you need to move this before `FactoryBot.create(:fact_value, fact_name: ansible_kernel_fact....`. If that fact exists, the RHSM fact won't be used.

However, now that I think about it, the RHSM fact is more likely to be correct since it checks in regularly while Ansible doesn't run on a regular basis.

@ares thoughts?

Reply:
Ideally, there should be a separate test for every single version.

Reply:
the correct way would have been adding this to reported_data_facet, however it behaves correctly now I'd say